File: algorithm/python/prob1001-1500/prob1245_medium_树的直径.py
class Solution:
    def treeDiameter(self, edges: List[List[int]]) -> int:
        n = len(edges)
        print(n)
        self.g = [[] for i in range(n + 1)]
        for e in edges:
            self.g[e[0]].append(e[1])
            self.g[e[1]].append(e[0])
        self.ans = 0
        self.dfs(0, -1)
        return self.ans

    def dfs(self, u, fa):
        max1 = max2 = 0
        for v in self.g[u]:
            if v != fa:
                t = self.dfs(v, u) + 1
                if max1 < t:
                    max2 = max1
                    max1 = t
                elif max2 < t:
                    max2 = t
        self.ans = max(self.ans, (max1 + max2))
        return max1

File: algorithm/java/prob1001-1500/prob1245_medium_树的直径.java
class Solution1 {
    private int ans;

    public int treeDiameter(int[][] edges) {
        int n = edges.length;
        ArrayList<Integer>[] g = new ArrayList[n + 1];
        for(int i = 0; i <= n; i++) {
            g[i] = new ArrayList<>();
        }
        for(int[] edge: edges) {
            g[edge[0]].add(edge[1]);
            g[edge[1]].add(edge[0]);
        }
        ans = 0;
        dfs(0, -1, g);
        return ans;
    }
    private int dfs(int u, int fa, ArrayList<Integer>[] g) {
        int max1 = 0;
        int max2 = 0;
        for(int v: g[u]) {
            if(v != fa) {
                int t = dfs(v, u, g) + 1;
                if(max1 < t) {
                    max2 = max1;
                    max1 = t;
                } else if (max2 < t){
                    max2 = t;
                }
            }
        }
        ans = Math.max(ans, (max1 + max2));
        return max1;
    }
}


class Solution {
    private int x; // 最远的点 x
    private int max_len; // 最长距离

    public int treeDiameter(int[][] edges) {
        int n = edges.length;
        ArrayList<Integer>[] g = new ArrayList[n + 1];
        for(int i = 0; i <= n; i++) {
            g[i] = new ArrayList<>();
        }
        for(int[] edge: edges) {
            g[edge[0]].add(edge[1]);
            g[edge[1]].add(edge[0]);
        }
        max_len = 0;
        x = 0;
        dfs(0, -1, 0, g);
        dfs(x, -1, 0, g);
        return max_len;
    }
    private void dfs(int u, int fa, int d, ArrayList<Integer>[] g) {
        if(d > max_len) {
            max_len = d;
            x = u;
        }
        for(int v: g[u]) {
            if(v != fa) {
                dfs(v, u, d + 1, g);
            }
        }
    }
}
